File: mlsauce/adaopt/adaopt.py
import numpy as np
import pickle
from sklearn.base import BaseEstimator
from sklearn.base import ClassifierMixin
from ..utils import subsample
from ..adaopt_cython import fit_adaopt, predict_proba_adaopt



class AdaOpt(BaseEstimator, ClassifierMixin):
    """AdaOpt classifier
        
       Attributes
       ----------
       n_iterations: int
           number of iterations of the optimizer at training time
       learning_rate: float
           controls the speed of the optimizer at training time  
       reg_lambda: float
           L2 regularization parameter for successive errors in the optimizer 
           (at training time)
       reg_alpha: float
           L1 regularization parameter for successive errors in the optimizer 
           (at training time)
       eta: float
           controls the slope in gradient descent (at training time)
       gamma: float
           controls the step size in gradient descent (at training time)
       k: int
           number of nearest neighbors selected at test time for classification
       tolerance: float
           controls early stopping in gradient descent (at training time)  
       n_clusters: int
            number of clusters, if MiniBatch k-means is used at test time 
            (for faster prediction)
       batch_size: int
            size of the batch, if MiniBatch k-means is used at test time 
            (for faster prediction)
       row_sample: float
           percentage of rows chosen from training set (by stratified subsampling, 
           for faster prediction)
       type_dist: str
           distance used for finding the nearest neighbors; currently `euclidean-f`
           (euclidean distances calculated as whole), `euclidean` (euclidean distances 
           calculated row by row), `cosine` (cosine distance)
       cache: boolean
           if the nearest neighbors are cached or not, for faster retrieval in 
           subsequent calls 
       seed: int 
           reproducibility seed for nodes_sim=='uniform', clustering and dropout
    """
    
    def __init__(
        self,
        n_iterations=50,
        learning_rate=0.3,
        reg_lambda=0.1,
        reg_alpha=0.5,
        eta=0.01,
        gamma=0.01,
        k=3,
        tolerance=0,
        n_clusters=0,
        batch_size=100,
        row_sample=0.8, 
        type_dist="euclidean-f",
        cache=True,
        seed=123,
    ):

        assert type_dist in (
            "euclidean",
            "euclidean-f",
            "cosine",
        ), "must have: `type_dist` in ('euclidean', 'euclidean-f', 'cosine') "

        self.n_iterations = n_iterations
        self.learning_rate = learning_rate
        self.reg_lambda = reg_lambda
        self.reg_alpha = reg_alpha
        self.eta = eta
        self.gamma = gamma
        self.k = k
        self.tolerance = tolerance
        self.n_clusters = n_clusters
        self.batch_size = batch_size
        self.row_sample = row_sample        
        self.type_dist = type_dist
        self.cache = cache
        self.seed = seed


    def fit(self, X, y, **kwargs):
        """Fit AdaOpt to training data (X, y)
        
        Parameters
        ----------
        X: {array-like}, shape = [n_samples, n_features]
            Training vectors, where n_samples is the number 
            of samples and n_features is the number of features
        
        y: array-like, shape = [n_samples]
               Target values
    
        **kwargs: additional parameters to be passed to self.cook_training_set
               
        Returns
        -------
        self: object
        """
        
        if self.row_sample < 1:
            index_subsample = subsample(y, row_sample=self.row_sample, 
                                       seed=self.seed)
            y_ = y[index_subsample]
            X_ = X[index_subsample, :]
        else:
            y_ = pickle.loads(pickle.dumps(y, -1))
            X_ = pickle.loads(pickle.dumps(X, -1))
        
        n, p = X_.shape
            
        n_classes = len(np.unique(y_))

        assert n == len(y_), "must have X.shape[0] == len(y)"

        res = fit_adaopt(
            X=X_,
            y=y_,
            n_iterations=self.n_iterations,
            n_X=n,
            p_X=p,
            n_classes=n_classes,
            learning_rate=self.learning_rate,
            reg_lambda=self.reg_lambda,
            reg_alpha=self.reg_alpha,
            eta=self.eta,
            gamma=self.gamma,
            tolerance=self.tolerance,
        )

        self.probs_training = res["probs"]
        self.training_accuracy = res["training_accuracy"]
        self.alphas = res["alphas"]
        self.n_iterations = res["n_iterations"]
        self.scaled_X_train = res["scaled_X_train"]

        return self


    def predict(self, X, **kwargs):
        """Predict test data X.
        
        Parameters
        ----------
        X: {array-like}, shape = [n_samples, n_features]
            Training vectors, where n_samples is the number 
            of samples and n_features is the number of features.
        
        **kwargs: additional parameters to be passed to `predict_proba`
                
               
        Returns
        -------
        model predictions: {array-like}
        """

        return np.argmax(self.predict_proba(X, **kwargs), 
                         axis=1)


    def predict_proba(self, X, **kwargs):
        """Predict probabilities for test data X.
        
        Parameters
        ----------
        X: {array-like}, shape = [n_samples, n_features]
            Training vectors, where n_samples is the number 
            of samples and n_features is the number of features.
        
        **kwargs: additional parameters to be passed to 
                  self.cook_test_set
               
        Returns
        -------
        probability estimates for test data: {array-like}        
        """

        n_train = self.scaled_X_train.shape[0]

        n_test = X.shape[0]
        
        return predict_proba_adaopt(X_test=X, 
                                scaled_X_train=self.scaled_X_train,
                                n_test=n_test, n_train=n_train,
                                probs_train=self.probs_training,
                                k=self.k, n_clusters=self.n_clusters,
                                batch_size=self.batch_size, 
                                type_dist=self.type_dist, 
                                cache=self.cache,
                                seed=self.seed)

File: R-package/R/adaopt.R



#' AdaOpt classifier
#'
#' @param n_iterations number of iterations of the optimizer at training time
#' @param learning_rate controls the speed of the optimizer at training time
#' @param reg_lambda L2 regularization parameter for successive errors in the optimizer (at training time)
#' @param reg_alpha L1 regularization parameter for successive errors in the optimizer (at training time)
#' @param eta controls the slope in gradient descent (at training time)
#' @param gamma controls the step size in gradient descent (at training time)
#' @param k number of nearest neighbors selected at test time for classification
#' @param tolerance controls early stopping in gradient descent (at training time)
#' @param n_clusters number of clusters, if MiniBatch k-means is used at test time (for faster prediction)
#' @param batch_size size of the batch, if MiniBatch k-means is used at test time (for faster prediction)
#' @param row_sample percentage of rows chosen from training set (by stratified subsampling, for faster prediction)
#' @param type_dist distance used for finding the nearest neighbors; currently `euclidean-f` (euclidean distances
#' calculated as whole), `euclidean` (euclidean distances calculated row by row), `cosine` (cosine distance)
#' @param cache if the nearest neighbors are cached or not, for faster retrieval in subsequent calls
#' @param seed reproducibility seed for initial weak learner and clustering
#'
#' @return An object of class AdaOpt
#' @export
#'
#' @examples
#'
#' library(datasets)
#'
#' X <- as.matrix(iris[, 1:4])
#' y <- as.integer(iris[, 5]) - 1L
#'
#' n <- dim(X)[1]
#' p <- dim(X)[2]
#' set.seed(21341)
#' train_index <- sample(x = 1:n, size = floor(0.8*n), replace = TRUE)
#' test_index <- -train_index
#' X_train <- as.matrix(iris[train_index, 1:4])
#' y_train <- as.integer(iris[train_index, 5]) - 1L
#' X_test <- as.matrix(iris[test_index, 1:4])
#' y_test <- as.integer(iris[test_index, 5]) - 1L
#'
#' obj <- AdaOpt()
#'
#' obj$fit(X_train, y_train)
#'
#' print(obj$score(X_test, y_test))
#'
AdaOpt <- function(n_iterations=50L,
                  learning_rate=0.3,
                  reg_lambda=0.1,
                  reg_alpha=0.5,
                  eta=0.01,
                  gamma=0.01,
                  k=3L,
                  tolerance=0,
                  n_clusters=0,
                  batch_size=100L,
                  row_sample=1.0,
                  type_dist="euclidean-f",
                  cache=TRUE,
                  seed=123L)
{

  ms$AdaOpt(n_iterations=n_iterations,
            learning_rate=learning_rate,
            reg_lambda=reg_lambda,
            reg_alpha=reg_alpha,
            eta=eta,
            gamma=gamma,
            k=k,
            tolerance=tolerance,
            n_clusters=n_clusters,
            batch_size=batch_size,
            row_sample=row_sample,
            type_dist=type_dist,
            cache=cache,
            seed=seed)
}
